modules/gmc-agent/src/main/java/org/gridgain/service/tracing/TracingService.java

```diff
@@ -54,39 +61,39 @@
     public TracingService(GridKernalContext ctx, WebSocketManager mgr) {
         this.ctx = ctx;
         this.mgr = mgr;
+        this.log = ctx.log(TracingService.class);
+        this.worker = createSenderWorker();
 
         ctx.grid().message().localListen(TRACING_TOPIC, onNodeTraces);
-    }
 
-    /**
-     * Send buffered spans.
-     */
-    public void sendInitialState() {
-        send(new SpanList());
+        exSrvc.submit(worker);
     }
 
     /** {@inheritDoc} */
     @Override public void close() {
         ctx.grid().message().stopLocalListen(TRACING_TOPIC, onNodeTraces);
+        exSrvc.shutdown();
     }
 
     /**
      * @param uuid Uuid.
      * @param spans Spans.
      */
     boolean onNodeTraces(UUID uuid, Object spans) {
-        send((SpanList) spans);
+        worker.addToSendQueue((SpanBatch) spans);
 
         return true;
     }
 
     /**
-     * @param spans Spans.
+     * @return Sender which send messages from queue to gmc.
      */
-    private void send(SpanList spans) {
-        buf.addAll(spans.getList());
-
-        if (mgr.send(buildSaveSpanDest(ctx.cluster().get().id()), Lists.newArrayList(buf)))
-            buf.clear();
+    private RetryableSender<SpanBatch> createSenderWorker() {
+        return new RetryableSender<SpanBatch>(log, QUEUE_CAP) {
+            @Override protected void send(SpanBatch list) {
+                if (!mgr.send(buildSaveSpanDest(ctx.cluster().get().id()), list.getList()))
+                    throw new IgniteException("Message with spans doesn't send");
```

Comment:
"Message with spans doesn't send" -> "Failed to send message with spans"

Reply:
Fix.